### ISSUE TITLE
Slow/delayed switching between active/inactive state with certain Gtk+ themes

### ISSUE THREAD

**Issue by thomholwerda** [IBBoard/cawbird#204]
I've been experiencing this really odd bug where switching to and from Cawbird as the active window is slow/delayed - but only with certain Gtk+ themes. I can't find any rhyme or reason as to why only certain themes cause this behaviour. All other applications do not exhibit this behaviour. The best way to show it is a video so... [I made one](https://photos.google.com/share/AF1QipMfmBcZuH_HgaAfi1cfFd33nUunBPU4AiI65lgmFn0gW-zmCYOjfStudDw1E-Hr_g/photo/AF1QipNjqPb0-G4IyH9iD8EIbnixu1qa5tWYpyNXTO3b?key=YXZwZkdKWjFGWFF4bm10ODR4ZlBYYUN0MzRRbjV3). With my phone, since I wasn't sure a screenrecording would actually capture it.

The theme I'm currently using is [this one](https://www.gnome-look.org/p/1267246/), and a theme I know for sure that doesn't have this issue is any of the many available Arc themes. The issue has been there forever as far I can remember, and while not a major bug, it is annoying as it also seems to cause the app to stutter a bit when switching to it.

This is on Linux Mint 20, Cinnamon, fully updated, running Cawbird 1.2.1. I think it may perhaps be related to a combination of Cinnamon's Mutter fork (Muffin) and Cawbird using CSD, but that's just a wild guess. Other CSD apps do NOT have this bug.

**Comment by IBBoard** [comment]
Is there a short (<1s) delay between you clicking to it and everything repainting?

I'm not sure how we debug this. I don't know of anything special that we do that might cause it. And I don't know how you'd debug something that's basically all GTK. I'll have to check some of the logging options.

**Comment by thomholwerda** [comment]
As you can hopefully see in the video, it doesn't seem it's a full redraw - it's the switch from the theme variant for inactive windows to the active one (and vice versa). To replicate, open e.g. a basic Gtk+ app (like Gedit or Xed) and Cawbird (with one of the offending themes set), switch between the two windows.

It could be some difference between how each theme is coded, but that still doesn't explain why only (as far as I can tell) Cawbird is affected. I've been trying to find another app that exhibits the issue, but no luck so far. I'm not a programmer, so this might be a dumb question, but perhaps Cawbird uses a rare Gtk+ element some themes stumble over?

Let me know if you need me to run a debug build or whatever, or anything else I can do to help!

**Comment by IBBoard** [comment]
It looks like I get it here as well, and just never noticed. I'm using a modified version of Adwaita, so it shouldn't have much affect other than having been loaded from CSS files in my home directory.

Oddly, if I bring up the Settings dialog then it is slow to swap between the main window and the Settings (or between the main window and another app) but quick between the Settings dialog and the other app. Which at least narrows it down to "probably the timeline widgets".

**Comment by IBBoard** [comment]
Well, it definitely appears to be `TweetListItem` related (i.e. the timeline). If I remove the Twitter API calls that make us populate the timelines then everything is quick and responsive.

It's not an image problem - if I disable image loading then it still has the lag.

It's not the fact that we've got scrollwindows inside scrollwindows (to stop tweets with very long unbreakable words stretching the whole window) - if I swap that out for a Bin then it's still slow.

If I add logging in to the `get_preferred_height_for_width` and `size_allocate` functions then it appears to do a lot of requesting new sizes for each widget, but as I understand it then that's just normal GTK behaviour. And we shouldn't do anything particularly expensive most of the time.

Anyone know how to debug GTK's repainting? 🙂

**Comment by thomholwerda** [comment]
Could it be an idea to file a Gtk+ bug? If this is out of your control, that seems like the right thing to do.

**Comment by IBBoard** [comment]
I want to rule out that we're not doing something odd that makes it self-inflicted first. There's nothing more annoying for libraries than getting bug reports that are just "I didn't use the API right". I'd probably go to the Gnome forums first.

**Comment by IBBoard** [comment]
More debugging: Even if we only have `HomeTimeline` available (remove the rest from the list in `MainWidget`) then it has the lag. And that's with ~40 items loaded.

**Comment by thomholwerda** [comment]
I don't have much useful stuff to add at this point, only to say that I am always surprised when something so small can lead to such a bug hunt :D.

**Comment by IBBoard** [comment]
I think that's called "software development" 😉

**Comment by IBBoard** [comment]
More debugging:
* If I quickly swap out the grids for a GtkBox then a) it looks ugly and b) it still lags (also, c) I broke something with avatar loading!)
* If I swap out the grids *and* just show the `HomeTimeline` then it's reasonably quick but not perfect
* If I keep the grids but just show the `HomeTimeline` then it's not showing as much lag as I saw previously

I'm starting to think that there's a problem between the "custom widget per tweet in a list" code and the way the UI handles the multiple views. I wonder whether @CodedOre is seeing any similar behaviour with #177 and the redesign?

**Comment by CodedOre** [comment]
I am not at that point to redesign the Tweet design in my libhandy-port (also took a break from working on that last week).

So, can't really give an experience here.

However, I have worked with an similar solution while reworking the AboutDialog (as I made a custom adaptive one, don't really know why).
I also had there a Grid which would basically be a list in adaptive view.

What I have done was to keep the grid, but to remove the elements from the grid and readd them in different positions when the change is done.

For one grid, it did worked well, but I can't say how the performance of that method would be when used to display the tweet.

**Comment by IBBoard** [comment]
I *think* this is separate from the 'responsive' design issue and moving widgets. And I only rearrange the grid if/when we cross the "responsive threshold", otherwise it just calls standard GTK code. I'd always assumed that removing and re-adding items would be slower than just adjusting a few attachment values, though!

I just grabbed a copy of your repo and tested out the main (only!) branch. It looks like you've still got the same issue, but then you're still using the same `MainWidget` with all the views, so it makes sense.

**Comment by CodedOre** [comment]
Well, like mentioned, I only implemented a similar system in the AboutDialog.

I currently looking trough the backend code to what and how I need to change it (as I need a few additional functionality for displaying a thread in the future), so I did not went to implementing more of the new UI yet.

**Comment by IBBoard** [comment]
Finally got round to asking on the Gnome forums - https://discourse.gnome.org/t/slow-active-inactive-state-change/5015

**Comment by IBBoard** [comment]
The official answer appears to be "we invalidate widgets; it happens".

But I'm wondering whether Cawbird's current design keeps all of the lists active, even when they're not visible, and so maybe we're invalidating more than just the current page? Maybe libhandy will improve that?

**Comment by IBBoard** [comment]
Reading linked tickets, [this one](https://gitlab.gnome.org/GNOME/gtk/-/issues/2302) suggests that `GtkStack` is the thing that performs badly. Which probably explains a lot, since all of the TweetList items are a stack to provide the "right-click to show actions" behaviour.

The libhandy mockups show actions in the TweetList (which takes more space, but is more accessible) so hopefully it won't suffer from most of these issues.

**Comment by IBBoard** [comment]
Those two commits seem to speed things up a bit BUT we lose the transitions. I don't think it's so important for actions, but I'm trying to keep the page transitions.

I've got something half-working with each page being in a `GtkRevealer` but I'm struggling to get the "tweet to tweet" transition working.

**Comment by IBBoard** [comment]
@thomholwerda Can you test the new branches? I was working on `bug204-slow-focus-painting` and it works and is quick, but the transitions aren't right yet. But I've just pushed `bug204-slow-focus-painting-alt` which was a random idea I had and I *think* it works and does the right thing.

**Comment by IBBoard** [comment]
I was going to reply to my original Gnome post to reference the other bug from Baedert and point out that there's a possible work-around… but it got auto-locked yesterday!

Hopefully people can still trip over this with Google. To increase the chances, here's my post text:

> One of my users has reported that my GTK3 app is slow to transition between “active” and “inactive” states. They’ve provided a video:
> 
> https://photos.google.com/share/AF1QipMfmBcZuH_HgaAfi1cfFd33nUunBPU4AiI65lgmFn0gW-zmCYOjfStudDw1E-Hr_g/photo/AF1QipNjqPb0-G4IyH9iD8EIbnixu1qa5tWYpyNXTO3b?key=YXZwZkdKWjFGWFF4bm10ODR4ZlBYYUN0MzRRbjV3 3
> 
> The delay is between the window getting focus and the window rendering in the “active” styling. You can click on the window, it’ll come to the front, there will be a short but noticeable delay, and then it’ll render as “active”. This happens with all themes, including Adwaita, so it’s not just the user’s theme causing it. It doesn’t happen when you’ve got a dialog (e.g. Settings) open.
> 
> The main window is a set of GtkListBox widgets containing custom GtkListItems that contain a GtkGrid with text and some image widgets.
> 
> If I limit it to 1 item in the list then the lag isn’t noticeable, but I wouldn’t expect 25 items (the default amount) to cause an obvious lag. It doesn’t appear to be the media, because the lag still happens when they’re not included.
> 
> What can cause the lag? How can I debug it? Is it just an unavoidable part of having lists with custom widgets in? From the logging I’ve added so far, it doesn’t appear to be a cascade of reallocations or anything, and I can’t find a way to debug CSS rendering.

**Comment by IBBoard** [comment]
I've done a bit more testing. Entirely removing the stack is definitely faster, but the animations are… substandard. Hiding non-visible things *seems* to be faster, but I've not got numbers to confirm.

**Comment by IBBoard** [comment]
Still working on fixing the self-to-self transitions, but there's now an "alt-alt" branch that goes even harder on the "remove things we don't need from the stack to speed up transitions". It required a few deeper changes, though, so it needs some testing before we make it the standard behaviour.

**Comment by IBBoard** [comment]
Has anyone been able to test the "alt-alt" branch yet?

**Comment by thomholwerda** [comment]
I want to, but don't know how, and didn't want to litter the thread 
with useless comments. However, since no one has replied, I figured I'd 
reply now!

**Comment by CodedOre** [comment]
> I want to, but don't know how, and didn't want to litter the thread 
> with useless comments. However, since no one has replied, I figured I'd 
> reply now!

It is not wrong to ask, if you ask me.

If you want to test the `alt-alt` branch, you need to clone the repository, use `git checkout` to switch to the branch and then build Cawbird to test it out.

**Comment by IBBoard** [comment]
How you build it partly depends on what distro you're on and how much of the tooling and libraries you've already got installed.

First you need the code:
```
git clone https://github.com/IBBoard/cawbird.git
```
Then you need the right branch:
```
git checkout bug204-slow-focus-painting-alt-alt
```
Then you need to [configure and compile it](https://github.com/IBBoard/cawbird#compiling):
```
meson build
ninja -C build
```
If that fails because it can't find the dependencies then you need to make sure that you've got the "…-devel" (or "…-dev" if you're on a Deb-based system) packages for [all of the dependencies](https://github.com/IBBoard/cawbird#build-dependencies).
And then you run it:
```
G_MESSAGE_DEBUG=cawbird build/cawbird
```

**Comment by IBBoard** [comment]
Any feedback on the `alt-alt` branch?

I've got to check whether I'm using the patches, but the bit that I tested it didn't seem to show any problems.

**Comment by IBBoard** [comment]
I've just rebased the `alt-alt` branch on top of Master, since it was quite out of date.

**Comment by IBBoard** [comment]
I *think* there's a bug in this. I've got a custom build that includes lots of new changes including this branch. When you right-click, it jumps to the top of the timeline. I presume something that I'm doing here is breaking the vadjust handling.

**Comment by IBBoard** [comment]
Okay, *now* it should be okay for people to test. I might still put some small improvements in, though.

**Comment by IBBoard** [comment]
I've been running with this as a custom patch for a little while and I've not noticed any problems, so I'll merge it.

**Comment by IBBoard** [comment]
Fixed by #335

**Comment by ValdikSS** [comment]
@IBBoard, unfortunately I can't say this is fixed for me in 1.5.
It may be worse earlier, but with GTK3 HighContrast theme, it redraws the window at 4 fps when it is focused or unfocused.
The issue is "backdrop" effect. Everything is very smooth with "Clearlooks-Phenix" theme which does not have it.

https://user-images.githubusercontent.com/3054729/174463549-a690f3d8-6770-44da-ae84-8d7d40c6b322.mp4

**Comment by IBBoard** [comment]
There's not really much more that we can do. We don't add backdrop effects, they're part of the system theme. We've already stripped back the unnecessary widgets (pages are removed when they're not the focussed one). I can't think of any other changes that we can make.

GTK4 apparently has better handling of lists of lots of widgets, so it will have to be something that is resolved with the redesign.
